contracts/BuilderStaking.sol

```diff
@@ -0,0 +1,330 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import {AccessControlDefaultAdminRulesUpgradeable} from '@openzeppelin/contracts-upgradeable/access/extensions/AccessControlDefaultAdminRulesUpgradeable.sol';
+import {UUPSUpgradeable} from '@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol';
+import {Initializable} from '@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol';
+import {SafeERC20} from '@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol';
+import {IERC20} from '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+
+/**
+ * @title Divvi BuilderStaking
+ * @dev Contract for managing $DIVVI token stakes on behalf of builders
+ * @custom:security-contact security@valora.xyz
+ */
+contract BuilderStaking is
+  Initializable,
+  AccessControlDefaultAdminRulesUpgradeable,
+  UUPSUpgradeable
+{
+  using SafeERC20 for IERC20;
+
+  // State variables
+  IERC20 public divviToken;
+  uint256 public stakingThreshold;
+  uint256 public totalStaked;
+
+  mapping(address => mapping(address => uint256)) public stakersForBeneficiary; // beneficiary => staker => amount
+  mapping(address => mapping(address => uint256)) public beneficiariesForStaker; // staker => beneficiary => amount
+
+  // Arrays to track relationships for efficient querying
+  mapping(address => address[]) public stakerListForBeneficiary;
+  mapping(address => address[]) public beneficiaryListForStaker;
```

Comment:
How many beneficiaries/stakers before looping over these arrays takes too much gas?
100 is fine, but what about 1,000 and then 10,000?

Should we reconsider storing these or use another data structure?

Reply:
I was considering this as well. We can of course just rely on indexing to fetch this data, which is imo a more typical approach, but means that the in order to render pertinent information about staking in the UI, we'll need to do that indexer work. That shouldn't be too difficult however, so it seems like we can probably just scrap this bookkeeping and rely on indexing to keep track of this.